pkg/kubelet/kubelet_pods.go

```diff
@@ -1138,28 +1138,35 @@ func (kl *Kubelet) HandlePodCleanups(ctx context.Context) error {
 		return err
 	}
 	for _, runningPod := range runningRuntimePods {
-		switch workerState, ok := workingPods[runningPod.ID]; {
-		case ok && workerState == SyncPod, ok && workerState == TerminatingPod:
-			// if the pod worker is already in charge of this pod, we don't need to do anything
-			continue
-		default:
-			// If the pod isn't in the set that should be running and isn't already terminating, terminate
-			// now. This termination is aggressive because all known pods should already be in a known state
-			// (i.e. a removed static pod should already be terminating), so these are pods that were
-			// orphaned due to kubelet restart or bugs. Since housekeeping blocks other config changes, we
-			// know that another pod wasn't started in the background so we are safe to terminate the
-			// unknown pods.
-			if _, ok := allPodsByUID[runningPod.ID]; !ok {
-				klog.V(3).InfoS("Clean up orphaned pod containers", "podUID", runningPod.ID)
+		// If the pod isn't in the set that should be running and isn't already terminating, terminate
+		// now. Since housekeeping is synchronous to other pod worker updates, we know that no pods have
+		// been added to the pod worker in the meantime. Note that pods that are not visible in the runtime
+		// but which were previously known are terminated by SyncKnownPods.
+		_, desiredPod := allPodsByUID[runningPod.ID]
+		knownPodState, knownPod := workingPods[runningPod.ID]
+
+		unknownPod, undesiredKnownSyncingPod := !knownPod, !desiredPod && knownPod && knownPodState == SyncPod
```

Comment:
Two types of unknown pods (maybe I can rename the variables, not sure whether that helps):

* Known to the pod worker (we started it)
* Known to the runtime (someone else besides this current pod worker started it)

I think (now that you are forcing me to go through this in my head), this loop could be reduced to having the responsibility only for unknown pods, which is really about telling the pod worker at most one time that a runningPod is lurking from some previous run and we should burn it with fire.  Once we tell the worker, it's now known to the pod worker, so we should skip it in the next HandlePodCleanups.  Originally, this loop was all about this special logic, but that's SyncKnownPods now (esp. the effective grace period calculation)

The one remaining concern would be that my logic around runningPod preservation is wrong for replay in lastUpdate - I'll add a unit test for that and simulate two separate HandlePodCleanups.

One final concern - once we have context cancellation, we need to make sure that once your grace period (as measured by the clock of the kubelet) has expired, you are hard terminated (we tell CRI to nuke you from orbit) and should need at most one more syncTerminatingPod loop.  We probably also need more insight into unknown pods via metrics - updating a metric once per loop of the number of unknown pods and the amount of time they expect to take to terminate or something.  We should also metric pod termination times I think if we can afford the cardinality.

Reply:
Thanks for the explanation! 

If we can reduce the cleanup here to just the pods unknown to the pod worker and have the pods known by pod worker handled by `SyncKnownPods`, it sounds like it will simplify the logic here, so +1 on making that change.

Regarding the context cancellation, it makes sense when we add that, but it sounds like we will need that change unrelated to `HandlePodCleanup` and it would responsibility of a different loop, is it correct?

Reply:
(also looking this over again I don't think we have a node e2e that ensures that unknown pods are terminated / cleaned up in general... let me add one to ensure this is tested in e2e).

I'm thinking something like:

1. create static pod or api pod
2. stop kubelet
3. delete the stop pod manifest or force delete the api pod
4. ensure that the pod is terminated by checking with CRI the underlying containers are removed

WDYT?

Reply:
have a WIP test here: https://github.com/kubernetes/kubernetes/pull/115442

Reply:
> Regarding the context cancellation, it makes sense when we add that, but it sounds like we will need that change unrelated to HandlePodCleanup and it would responsibility of a different loop, is it correct?

Would be a responsibility of the methods that shorten grace period, probably a responsibility of the caller of syncTerminatingPod (pod worker loop) to calculate the allowed grace period based on wall clock, and then maybe even a change further down in CRI to treat "really kill" differently.  At minimum I found a bug in container runtime where "force delete" and "grace delete" code paths are not correctly paired to what pod worker does (pod worker should pass 0 in those cases, not non-positive, and container runtime should treat zero as bypassing preStop and various other niceties).

Will review #115422, we probably want to test both api pod and static pod if we possibly can because of the number of places with `if static pod` logic.

Reply:
Thanks, do you have a bit more context on the issue with force deletion / grace period not being set correctly? It would be nice if we can file issue to ensure we can keep track of the fixes.

> Will review #115422, we probably want to test both api pod and static pod if we possibly can because of the number of places with if static pod logic.

Thanks, let me know if you have any feedback there. I included both cases, api pod and static pod since as you mentioned we have some special handling for static pods.